File: shared/services/trading_service.py
"""
Trading Service - Python implementation
Client for Python trading service integration
"""

import asyncio
import aiohttp
import json
from typing import List, Dict, Optional, Any
from datetime import datetime

from ..api.base_api_client import ExtendedApiClient
from ..types.trading_types import (
    OrderSide, OrderType, OrderStatus, TradingPair, 
    OrderRequest, OrderResponse, MarketData, Balance
)
from ..errors.app_error import TradingError, ErrorCodes
from ..logging.logger_service import LoggerService


class TradingService(ExtendedApiClient):
    """Python Trading Service Client"""
    
    def __init__(self, base_url: str = "http://localhost:8000"):
        super().__init__(base_url)
        self.logger = LoggerService()
    
    async def get_markets(self) -> List[TradingPair]:
        """Get available markets"""
        try:
            response = await self.get("/markets")
            
            if response.status == 200:
                markets_data = await response.json()
                markets = []
                
                for market_json in markets_data:
                    market = TradingPair(
                        base_asset=market_json.get('base_asset', ''),
                        quote_asset=market_json.get('quote_asset', ''),
                        symbol=market_json.get('symbol', ''),
                        min_quantity=float(market_json.get('min_qty', '0')),
                        max_quantity=float(market_json.get('max_qty', '0')),
                        price_precision=8,
                        quantity_precision=8,
                        is_active=market_json.get('status') == 'ACTIVE'
                    )
                    markets.append(market)
                
                return markets
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to fetch markets: {response.status}"
            )
            
        except Exception as e:
            self.logger.error(
                "Failed to get markets",
                metadata={"error": str(e)},
                service="trading_service"
            )
            
            if isinstance(e, TradingError):
                raise e
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to get markets: {e}"
            )
    
    async def get_market_data(self, symbol: str) -> MarketData:
        """Get market data for specific symbol"""
        try:
            response = await self.get(f"/markets/{symbol}/data")
            
            if response.status == 200:
                data_json = await response.json()
                
                return MarketData(
                    symbol=data_json.get('symbol', symbol),
                    price=float(data_json.get('price', '0')),
                    change_24h=float(data_json.get('change_24h', '0')),
                    volume_24h=float(data_json.get('volume_24h', '0')),
                    high_24h=float(data_json.get('high_24h', '0')),
                    low_24h=float(data_json.get('low_24h', '0')),
                    timestamp=datetime.now()
                )
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to fetch market data: {response.status}"
            )
            
        except Exception as e:
            self.logger.error(
                "Failed to get market data",
                metadata={"symbol": symbol, "error": str(e)},
                service="trading_service"
            )
            
            if isinstance(e, TradingError):
                raise e
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to get market data: {e}"
            )
    
    async def get_account_balance(self) -> Balance:
        """Get account balance"""
        try:
            response = await self.get("/account/balance")
            
            if response.status == 200:
                balance_json = await response.json()
                
                return Balance(
                    total_balance=float(balance_json.get('total_balance', '0')),
                    available_balance=float(balance_json.get('available_balance', '0')),
                    reserved_balance=float(balance_json.get('margin_used', '0')),
                    currency=balance_json.get('currency', 'USD')
                )
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to fetch balance: {response.status}"
            )
            
        except Exception as e:
            self.logger.error(
                "Failed to get account balance",
                metadata={"error": str(e)},
                service="trading_service"
            )
            
            if isinstance(e, TradingError):
                raise e
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to get account balance: {e}"
            )
    
    async def create_order(self, order_request: OrderRequest) -> OrderResponse:
        """Create trading order"""
        try:
            order_json = {
                'market': order_request.symbol,
                'side': order_request.side.value.upper(),
                'qty': str(order_request.quantity),
                'price': str(order_request.price),
                'order_type': order_request.type.value.upper(),
                'time_in_force': 'GTC',
                'reduce_only': False,
            }
            
            response = await self.post("/orders", json=order_json)
            
            if response.status == 200:
                response_json = await response.json()
                
                return OrderResponse(
                    order_id=response_json.get('order_id', ''),
                    symbol=order_request.symbol,
                    side=order_request.side,
                    quantity=order_request.quantity,
                    price=order_request.price,
                    status=OrderStatus.CREATED if response_json.get('status') == 'CREATED' else OrderStatus.FAILED,
                    message=response_json.get('message', ''),
                    timestamp=datetime.now()
                )
            
            # Handle error response
            if response.status == 400:
                error_json = await response.json()
                error_message = error_json.get('detail', 'Order creation failed')
                raise TradingError(
                    code=ErrorCodes.TRADING_INVALID_ORDER,
                    message=error_message
                )
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to create order: {response.status}"
            )
            
        except Exception as e:
            self.logger.error(
                "Failed to create order",
                metadata={
                    "symbol": order_request.symbol,
                    "side": order_request.side.value,
                    "quantity": str(order_request.quantity),
                    "price": str(order_request.price),
                    "error": str(e),
                },
                service="trading_service"
            )
            
            if isinstance(e, TradingError):
                raise e
            
            raise TradingError(
                code=ErrorCodes.TRADING_API_ERROR,
                message=f"Failed to create order: {e}"
            )
    
    async def test_connection(self) -> bool:
        """Test connection to trading service"""
        try:
            response = await self.get("/test/connection")
            
            if response.status == 200:
                connection_json = await response.json()
                return connection_json.get('status') == 'success'
            
            return False
            
        except Exception as e:
            self.logger.error(
                "Trading service connection test failed",
                metadata={"error": str(e)},
                service="trading_service"
            )
            return False
    
    async def get_health_status(self) -> Dict[str, Any]:
        """Get service health status"""
        try:
            response = await self.get("/health")
            
            if response.status == 200:
                return await response.json()
            
            return {'status': 'unhealthy', 'error': f'HTTP {response.status}'}
            
        except Exception as e:
            return {'status': 'unhealthy', 'error': str(e)}
    
    async def execute_trade(
        self,
        symbol: str,
        side: OrderSide,
        quantity: float,
        price: float,
        order_type: OrderType = OrderType.LIMIT
    ) -> OrderResponse:
        """Execute a trade with automatic signature generation"""
        
        self.logger.info(
            'Executing trade via Python service',
            metadata={
                'symbol': symbol,
                'side': side.value,
                'quantity': str(quantity),
                'price': str(price),
                'type': order_type.value,
            },
            service="trading_service"
        )
        
        order_request = OrderRequest(
            symbol=symbol,
            side=side,
            quantity=quantity,
            price=price,
            type=order_type,
        )
        
        return await self.create_order(order_request)


class TradingServiceConfig:
    """Trading Service Configuration"""
    
    DEFAULT_BASE_URL = "http://localhost:8000"
    DEFAULT_TIMEOUT = 30
    MAX_RETRIES = 3
    
    @staticmethod
    async def is_service_available(base_url: str = DEFAULT_BASE_URL) -> bool:
        """Check if trading service is available"""
        try:
            async with aiohttp.ClientSession(timeout=aiohttp.ClientTimeout(total=5)) as session:
                async with session.get(f"{base_url}/health") as response:
                    return response.status == 200
        except Exception:
            return False

File: shared/services/trading_service.dart
/// Trading Service
/// Connects to Python trading service for Extended Exchange integration
library;

import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import '../api/base_api_client.dart';
import '../types/trading_types.dart';
import '../errors/app_error.dart';
import '../logging/logger_service.dart';

/// Python Trading Service Client
class TradingService extends BaseApiClient {
  static const String _baseUrl = 'http://localhost:8000';
  
  TradingService() : super(_baseUrl);
  
  /// Get available markets
  Future<List<TradingPair>> getMarkets() async {
    try {
      final response = await get('/markets');
      
      if (response.statusCode == 200) {
        final List<dynamic> marketsJson = json.decode(response.body);
        return marketsJson.map((json) => TradingPair.fromJson(json)).toList();
      }
      
      throw TradingError.apiError('Failed to fetch markets: ${response.statusCode}');
    } catch (e) {
      LoggerService().error('Failed to get markets', metadata: {'error': e.toString()});
      throw TradingError.apiError('Failed to get markets: $e');
    }
  }
  
  /// Get market data for specific symbol
  Future<MarketData> getMarketData(String symbol) async {
    try {
      final response = await get('/markets/$symbol/data');
      
      if (response.statusCode == 200) {
        final Map<String, dynamic> dataJson = json.decode(response.body);
        return MarketData.fromJson(dataJson);
      }
      
      throw TradingError.apiError('Failed to fetch market data: ${response.statusCode}');
    } catch (e) {
      LoggerService().error('Failed to get market data', metadata: {'symbol': symbol, 'error': e.toString()});
      throw TradingError.apiError('Failed to get market data: $e');
    }
  }
  
  /// Get account balance
  Future<Balance> getAccountBalance() async {
    try {
      final response = await get('/account/balance');
      
      if (response.statusCode == 200) {
        final Map<String, dynamic> balanceJson = json.decode(response.body);
        return Balance.fromJson(balanceJson);
      }
      
      throw TradingError.apiError('Failed to fetch balance: ${response.statusCode}');
    } catch (e) {
      LoggerService().error('Failed to get account balance', metadata: {'error': e.toString()});
      throw TradingError.apiError('Failed to get account balance: $e');
    }
  }
  
  /// Create trading order
  Future<OrderResponse> createOrder(OrderRequest orderRequest) async {
    try {
      final orderJson = {
        'market': orderRequest.symbol,
        'side': orderRequest.side.name.toUpperCase(),
        'qty': orderRequest.quantity.toString(),
        'price': orderRequest.price.toString(),
        'order_type': orderRequest.type.name.toUpperCase(),
        'time_in_force': 'GTC',
        'reduce_only': false,
      };
      
      final response = await post('/orders', body: json.encode(orderJson));
      
      if (response.statusCode == 200) {
        final Map<String, dynamic> responseJson = json.decode(response.body);
        return OrderResponse.fromJson(responseJson);
      }
      
      // Handle error response
      if (response.statusCode == 400) {
        final Map<String, dynamic> errorJson = json.decode(response.body);
        final errorMessage = errorJson['detail'] ?? 'Order creation failed';
        throw TradingError.invalidOrder(errorMessage);
      }
      
      throw TradingError.apiError('Failed to create order: ${response.statusCode}');
    } catch (e) {
      LoggerService().error(
        'Failed to create order',
        metadata: {
          'symbol': orderRequest.symbol,
          'side': orderRequest.side.name,
          'quantity': orderRequest.quantity.toString(),
          'price': orderRequest.price.toString(),
          'error': e.toString(),
        },
      );
      
      if (e is TradingError) {
        rethrow;
      }
      
      throw TradingError.apiError('Failed to create order: $e');
    }
  }
  
  /// Test connection to trading service
  Future<bool> testConnection() async {
    try {
      final response = await get('/test/connection');
      
      if (response.statusCode == 200) {
        final Map<String, dynamic> connectionJson = json.decode(response.body);
        return connectionJson['status'] == 'success';
      }
      
      return false;
    } catch (e) {
      LoggerService().error('Trading service connection test failed', metadata: {'error': e.toString()});
      return false;
    }
  }
  
  /// Get service health status
  Future<Map<String, dynamic>> getHealthStatus() async {
    try {
      final response = await get('/health');
      
      if (response.statusCode == 200) {
        return json.decode(response.body);
      }
      
      return {'status': 'unhealthy', 'error': 'HTTP ${response.statusCode}'};
    } catch (e) {
      return {'status': 'unhealthy', 'error': e.toString()};
    }
  }
}

/// Trading Service Provider Extensions
extension TradingServiceExtensions on TradingService {
  /// Execute a trade with automatic signature generation
  Future<OrderResponse> executeTrade({
    required String symbol,
    required OrderSide side,
    required double quantity,
    required double price,
    OrderType type = OrderType.limit,
  }) async {
    LoggerService().info(
      'Executing trade via Python service',
      metadata: {
        'symbol': symbol,
        'side': side.name,
        'quantity': quantity.toString(),
        'price': price.toString(),
        'type': type.name,
      },
    );
    
    final orderRequest = OrderRequest(
      symbol: symbol,
      side: side,
      quantity: quantity,
      price: price,
      type: type,
    );
    
    return await createOrder(orderRequest);
  }
  
  /// Get real-time market data with caching
  Future<MarketData> getCachedMarketData(String symbol) async {
    // This could be enhanced with caching logic
    return await getMarketData(symbol);
  }
}

/// Trading Service Configuration
class TradingServiceConfig {
  static const String defaultBaseUrl = 'http://localhost:8000';
  static const Duration defaultTimeout = Duration(seconds: 30);
  static const int maxRetries = 3;
  
  /// Check if trading service is available
  static Future<bool> isServiceAvailable() async {
    try {
      final client = http.Client();
      final response = await client.get(
        Uri.parse('$defaultBaseUrl/health'),
        headers: {'Content-Type': 'application/json'},
      ).timeout(const Duration(seconds: 5));
      
      client.close();
      
      return response.statusCode == 200;
    } catch (e) {
      return false;
    }
  }
}